pint_xarray/checking.py

```diff
@@ -0,0 +1,186 @@
+        (or will attempt to convert the argument to these units), then will strip the units before passing the magnitude
+        to the wrapped function.
+
+        A value of None indicates not to check that argument for units (suitable for flags and other non-data
+        arguments).
+
+    Returns
+    -------
+    return_values : Any
+        Return values of the wrapped function, either a single value or a tuple of values. These will be given units
+        according to return_units.
+
+    Raises
+    ------
+    TypeError
+        If an argument or return value has a specified unit, but is not an xarray.DataArray or pint.Quantity.
+        Also thrown if any of the units are not a valid type, or if the number of arguments or return values does not
+        match the number of units specified.
+
+    Examples
+    --------
+
+    Decorating a function which takes one quantified input, but returns a non-data value (in this case a boolean).
+
+    >>> @expects("deg C")
+    ... def above_freezing(temp):
+    ...     return temp > 0
+
+
+    TODO: example where we check units of an optional weighted kwarg
+    """
+
+    # Check types of args_units, kwargs_units, and return_units
+    all_units = list(args_units) + list(kwargs_units.values())
+    if isinstance(return_units, list):
+        all_units = all_units + return_units
+    elif return_units:
+        all_units = all_units + [return_units]
+    for a in all_units:
+        if isinstance(a, dict):
+            for u in a.values():
+                _check_valid_unit_type(u)
+        else:
+            _check_valid_unit_type(a)
+
+    def _expects_decorator(func):
+        @functools.wraps(func)
+        def _unit_checking_wrapper(*args, **kwargs):
+
+            # check same number of arguments were passed as expected
+            if len(args) != len(args_units):
+                raise TypeError(
+                    f"{len(args)} arguments were passed, but {len(args_units)} arguments were expected"
+                )
+
+            converted_args = []
+            for arg, arg_unit in zip(args, args_units):
+                converted_arg = _check_or_convert_to_then_strip(arg, arg_unit)
+                converted_args.append(converted_arg)
+
+            converted_kwargs = {}
+            for key, val in kwargs.items():
+                kwarg_unit = kwargs_units.get(key, None)
+                converted_kwargs[key] = _check_or_convert_to_then_strip(val, kwarg_unit)
+
+            results = func(*converted_args, **converted_kwargs)
+
+            if results is None:
+                if return_units:
+                    raise TypeError(
+                        "Expected function to return something, but function returned None"
+                    )
+            else:
+                if return_units is None:
+                    # ignore types and units of return values
+                    return results
```

Comment:
> this will raise if return_units is not None and results is None

That seems better to me, and this change passes the tests, so I've added it in https://github.com/xarray-contrib/pint-xarray/pull/143/commits/4de6f4d66e534c8e769fdf874c778afff34b5196